File: mailblastr-php/src/Exceptions/MailblastrException.php
<?php

declare(strict_types=1);

namespace Mailblastr\Exceptions;

/**
 * Thrown for every non-2xx API response (and transport-level failures).
 *
 * Carries the API error shape: `statusCode`, `name`, `message`.
 *
 * Branch on {@see getName()} together with {@see getStatusCode()}, never on the
 * message: message text is scrubbed of provider identifiers server-side and is
 * not a stable contract, and a handler may answer with a status other than the
 * one a name usually maps to.
 *
 * Some errors are a superset of that envelope. The whole parsed body is kept on
 * {@see getBody()}, with the common extras surfaced as accessors that return
 * null on an ordinary error:
 *
 * - {@see getLimit()} — plan/quota rejections (`plan_limit_reached`,
 *   `*_quota_exceeded`, `contact_limit_reached`, `ai_credits_exceeded`) say
 *   WHICH cap was hit and what would clear it.
 * - {@see getReputation()} — reputation gates (`reputation_paused`,
 *   `reputation_limit_exceeded`).
 * - {@see getSent()} / {@see getSentCount()} — a `POST /emails/batch` that
 *   failed part way through, naming the emails that DID go out.
 */
class MailblastrException extends \Exception
{
    /**
     * @param array<string, mixed> $body The full parsed error body ([] when the
     *                                   response was not a JSON object).
     */
    public function __construct(
        string $message,
        private readonly int $statusCode = 0,
        private readonly string $errorName = 'application_error',
        private readonly array $body = [],
    ) {
        parent::__construct($message);
    }

    /**
     * Build an exception from a decoded API error body
     * (`{ statusCode, name, message }`), falling back to the HTTP status. The
     * whole body is retained, so the additive `limit` / `reputation` / `sent`
     * fields survive.
     *
     * @param array<string, mixed> $body
     */
    public static function fromResponse(int $httpStatus, array $body): self
    {
        $message = isset($body['message']) && is_string($body['message'])
            ? $body['message']
            : "Request failed with status {$httpStatus}";
        $statusCode = isset($body['statusCode']) && is_int($body['statusCode'])
            ? $body['statusCode']
            : $httpStatus;
        $name = isset($body['name']) && is_string($body['name'])
            ? $body['name']
            : 'application_error';

        return new self($message, $statusCode, $name, $body);
    }

    /** The HTTP status code (0 for network-level errors). */
    public function getStatusCode(): int
    {
        return $this->statusCode;
    }

    /** The machine-readable API error name, e.g. 'validation_error'. */
    public function getName(): string
    {
        return $this->errorName;
    }

    /**
     * The full parsed error body — [] when the response was not a JSON object.
     * Read it for any additive field newer than this SDK version.
     *
     * @return array<string, mixed>
     */
    public function getBody(): array
    {
        return $this->body;
    }

    /**
     * The plan/quota cap this request hit, or null when the error is not a
     * limit error. `kind` says WHICH quota ran out; `next_plan` is null when
     * only Enterprise would fit, and `credits` is present only for the
     * email-quota kinds.
     *
     * @return null|array{
     *     kind: string, used: int, limit: int, requested?: int, remaining?: int,
     *     period?: string, plan?: array{id: string, name: string},
     *     next_plan?: null|array{id: string, name: string, amount: int, currency: string,
     *                            monthly_emails: int, daily_emails: int, domains: int,
     *                            contacts: int, ai_credits: int, automation_runs: int},
     *     credits?: array{balance: int, needed: int, purchasable: bool,
     *                     unit: int, amount_per_unit_cents: int}
     * }
     */
    public function getLimit(): ?array
    {
        return $this->arrayField('limit');
    }

    /**
     * The reputation-gate detail on `reputation_paused` /
     * `reputation_limit_exceeded`, else null.
     *
     * @return null|array{
     *     retryable: bool, scope?: string, status?: string, scope_key?: string,
     *     hourly_limit?: int, daily_limit?: int, hourly_used?: int, daily_used?: int,
     *     retry_at?: string, support_email?: string
     * }
     */
    public function getReputation(): ?array
    {
        return $this->arrayField('reputation');
    }

    /**
     * The emails that were already sent before a batch failed part way through
     * (`POST /emails/batch` with an `idempotencyKey`), else null. Do NOT resend
     * these.
     *
     * @return null|list<array{id: string}>
     */
    public function getSent(): ?array
    {
        $value = $this->body['sent'] ?? null;

        return is_array($value) ? $value : null;
    }

    /**
     * How many emails went out before a batch failed part way through, else
     * null. Falls back to {@see getSent()}'s length when the body carried the
     * list but not the count.
     */
    public function getSentCount(): ?int
    {
        $count = $this->body['sent_count'] ?? null;
        if (is_int($count)) {
            return $count;
        }
        $sent = $this->getSent();

        return $sent === null ? null : count($sent);
    }

    /** @return null|array<string, mixed> */
    private function arrayField(string $key): ?array
    {
        $value = $this->body[$key] ?? null;

        return is_array($value) ? $value : null;
    }
}

File: mailblastr-dotnet/src/Mailblastr/MailblastrException.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Mailblastr;

/// <summary>
/// Thrown when the MailBlastr API returns a non-2xx response (or the response
/// cannot be reached/parsed). Carries the API error shape:
/// <c>{ statusCode, name, message }</c>.
/// </summary>
/// <remarks>
/// Some errors are a superset of that envelope. The additive fields are surfaced
/// as typed properties, and are <c>null</c> on an ordinary error:
/// <list type="bullet">
///   <item><see cref="Limit"/> — plan/quota rejections (<c>plan_limit_reached</c>,
///   <c>daily_quota_exceeded</c>, <c>monthly_quota_exceeded</c>,
///   <c>contact_limit_reached</c>, <c>ai_credits_exceeded</c>,
///   <c>automation_quota_exceeded</c>) say WHICH cap was hit.</item>
///   <item><see cref="Reputation"/> — reputation gates (<c>reputation_paused</c>,
///   <c>reputation_limit_exceeded</c>).</item>
///   <item><see cref="Sent"/> / <see cref="SentCount"/> — a
///   <c>POST /emails/batch</c> that failed part way through, naming the emails
///   that DID go out.</item>
/// </list>
/// Anything else the body carried stays reachable through <see cref="Extra"/>.
/// Branch on <see cref="Name"/> together with <see cref="StatusCode"/>, never on
/// <see cref="Exception.Message"/> — message text is scrubbed server-side and is
/// not a stable contract.
/// </remarks>
public class MailblastrException : Exception
{
    /// <summary>HTTP status code of the failed response (0 for network errors).</summary>
    public int StatusCode { get; }

    /// <summary>Machine-readable error name, e.g. <c>validation_error</c> or <c>not_found</c>.</summary>
    public string Name { get; }

    /// <summary>
    /// Every field the error body carried beyond <c>statusCode</c>, <c>name</c>
    /// and <c>message</c>, unparsed. Empty when the response was not a JSON
    /// object. Use it to read fields newer than this SDK version.
    /// </summary>
    public IReadOnlyDictionary<string, JsonElement> Extra { get; }

    /// <summary>
    /// The plan/quota cap this request hit, or <c>null</c> when the error is not
    /// a limit error. Tells you WHICH quota ran out (<c>Kind</c>), how much of it
    /// was used, and the cheapest plan that would fit.
    /// </summary>
    public PlanLimitDetail? Limit { get; }

    /// <summary>
    /// The reputation-gate detail on <c>reputation_paused</c> /
    /// <c>reputation_limit_exceeded</c>, else <c>null</c>.
    /// </summary>
    public ReputationDetail? Reputation { get; }

    /// <summary>
    /// The emails that were already sent before a batch failed part way through
    /// (<c>POST /emails/batch</c> with an <c>Idempotency-Key</c>), else
    /// <c>null</c>. Do NOT resend these.
    /// </summary>
    public IReadOnlyList<EmailCreated>? Sent { get; }

    /// <summary>
    /// How many emails went out before a batch failed part way through, else
    /// <c>null</c>. Falls back to <see cref="Sent"/>'s length when the body
    /// carried the list but not the count.
    /// </summary>
    public int? SentCount { get; }

    public MailblastrException(int statusCode, string name, string message)
        : this(statusCode, name, message, (IReadOnlyDictionary<string, JsonElement>?)null)
    {
    }

    public MailblastrException(int statusCode, string name, string message, Exception innerException)
        : base(message, innerException)
    {
        StatusCode = statusCode;
        Name = name;
        Extra = EmptyExtra;
    }

    /// <param name="statusCode">HTTP status of the failed response.</param>
    /// <param name="name">The error body's machine-readable <c>name</c>.</param>
    /// <param name="message">The error body's <c>message</c>.</param>
    /// <param name="extra">
    /// The additive fields of the parsed error body (everything except
    /// <c>statusCode</c> / <c>name</c> / <c>message</c>), or <c>null</c> when the
    /// response was not a JSON object.
    /// </param>
    public MailblastrException(int statusCode, string name, string message, IReadOnlyDictionary<string, JsonElement>? extra)
        : base(message)
    {
        StatusCode = statusCode;
        Name = name;
        Extra = extra ?? EmptyExtra;

        Limit = Read<PlanLimitDetail>(Extra, "limit");
        Reputation = Read<ReputationDetail>(Extra, "reputation");
        Sent = Read<List<EmailCreated>>(Extra, "sent");
        SentCount = ReadSentCount(Extra, Sent);
    }

    private static readonly IReadOnlyDictionary<string, JsonElement> EmptyExtra =
        new Dictionary<string, JsonElement>();

    /// <summary>
    /// Deserialize one additive field, tolerating a shape this SDK version does
    /// not know: a field we cannot parse becomes <c>null</c> rather than costing
    /// the caller the whole exception. The raw value is still in <see cref="Extra"/>.
    /// </summary>
    private static T? Read<T>(IReadOnlyDictionary<string, JsonElement> extra, string key) where T : class
    {
        if (!extra.TryGetValue(key, out var element) || element.ValueKind is JsonValueKind.Null or JsonValueKind.Undefined)
        {
            return null;
        }
        try
        {
            return element.Deserialize<T>(MailblastrJson.Options);
        }
        catch (JsonException)
        {
            return null;
        }
    }

    private static int? ReadSentCount(IReadOnlyDictionary<string, JsonElement> extra, IReadOnlyList<EmailCreated>? sent)
    {
        if (extra.TryGetValue("sent_count", out var element)
            && element.ValueKind == JsonValueKind.Number
            && element.TryGetInt32(out var count))
        {
            return count;
        }
        return sent?.Count;
    }
}

/// <summary>
/// The plan/quota cap attached to a limit rejection (see
/// <see cref="MailblastrException.Limit"/>).
/// </summary>
public class PlanLimitDetail
{
    /// <summary>
    /// Which cap was hit: <c>emails_daily</c>, <c>emails_monthly</c>,
    /// <c>domains</c>, <c>automation_runs</c>, <c>ai_credits</c>,
    /// <c>contacts</c> or <c>campaign_preflight</c>. Treat it as an open set.
    /// </summary>
    [JsonPropertyName("kind")]
    public string Kind { get; set; } = null!;

    /// <summary>How much of the allowance is already consumed.</summary>
    [JsonPropertyName("used")]
    public long Used { get; set; }

    /// <summary>The allowance itself.</summary>
    [JsonPropertyName("limit")]
    public long Limit { get; set; }

    /// <summary>How much this request asked for.</summary>
    [JsonPropertyName("requested")]
    public long? Requested { get; set; }

    /// <summary>What is left of the allowance.</summary>
    [JsonPropertyName("remaining")]
    public long? Remaining { get; set; }

    /// <summary>Rolling window the cap is measured over: <c>24h</c> or <c>30d</c>.</summary>
    [JsonPropertyName("period")]
    public string? Period { get; set; }

    /// <summary>The plan the account is on.</summary>
    [JsonPropertyName("plan")]
    public PlanRef? Plan { get; set; }

    /// <summary>The cheapest plan that would fit, or <c>null</c> when only Enterprise does.</summary>
    [JsonPropertyName("next_plan")]
    public NextPlanRef? NextPlan { get; set; }

    /// <summary>Top-up credits — present only for the email-quota kinds.</summary>
    [JsonPropertyName("credits")]
    public PlanLimitCredits? Credits { get; set; }
}

/// <summary>The account's current plan, as named on a limit error.</summary>
public class PlanRef
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;
}

/// <summary>The cheapest plan that would clear the cap, with its allowances.</summary>
public class NextPlanRef : PlanRef
{
    /// <summary>Price in the minor unit of <see cref="Currency"/> (e.g. cents).</summary>
    [JsonPropertyName("amount")]
    public int Amount { get; set; }

    [JsonPropertyName("currency")]
    public string Currency { get; set; } = null!;

    [JsonPropertyName("monthly_emails")]
    public long MonthlyEmails { get; set; }

    [JsonPropertyName("daily_emails")]
    public long DailyEmails { get; set; }

    [JsonPropertyName("domains")]
    public long Domains { get; set; }

    [JsonPropertyName("contacts")]
    public long Contacts { get; set; }

    [JsonPropertyName("ai_credits")]
    public long AiCredits { get; set; }

    [JsonPropertyName("automation_runs")]
    public long AutomationRuns { get; set; }
}

/// <summary>Top-up credit balance offered alongside an email-quota rejection.</summary>
public class PlanLimitCredits
{
    [JsonPropertyName("balance")]
    public long Balance { get; set; }

    [JsonPropertyName("needed")]
    public long Needed { get; set; }

    [JsonPropertyName("purchasable")]
    public bool Purchasable { get; set; }

    /// <summary>Credits sold per unit (e.g. 1000 emails).</summary>
    [JsonPropertyName("unit")]
    public long Unit { get; set; }

    [JsonPropertyName("amount_per_unit_cents")]
    public int AmountPerUnitCents { get; set; }
}

/// <summary>
/// The reputation-gate detail attached to <c>reputation_paused</c> /
/// <c>reputation_limit_exceeded</c> (see <see cref="MailblastrException.Reputation"/>).
/// </summary>
public class ReputationDetail
{
    /// <summary>Whether waiting and retrying can succeed.</summary>
    [JsonPropertyName("retryable")]
    public bool Retryable { get; set; }

    /// <summary>What the gate applies to: <c>tenant</c>, <c>domain</c> or <c>platform</c>.</summary>
    [JsonPropertyName("scope")]
    public string? Scope { get; set; }

    /// <summary>The reputation status that triggered the gate.</summary>
    [JsonPropertyName("status")]
    public string? Status { get; set; }

    /// <summary>The specific tenant/domain the scope refers to.</summary>
    [JsonPropertyName("scope_key")]
    public string? ScopeKey { get; set; }

    [JsonPropertyName("hourly_limit")]
    public long? HourlyLimit { get; set; }

    [JsonPropertyName("daily_limit")]
    public long? DailyLimit { get; set; }

    [JsonPropertyName("hourly_used")]
    public long? HourlyUsed { get; set; }

    [JsonPropertyName("daily_used")]
    public long? DailyUsed { get; set; }

    /// <summary>When sending may resume.</summary>
    [JsonPropertyName("retry_at")]
    public string? RetryAt { get; set; }

    [JsonPropertyName("support_email")]
    public string? SupportEmail { get; set; }

    /// <summary>Any reputation field newer than this SDK version.</summary>
    [JsonExtensionData]
    public Dictionary<string, JsonElement>? Extra { get; set; }
}
